Add basic completions for `nixos-option`

Ma27 opened the pull request:
The `nixos-option` script can be fairly helpful for option discovery on
the currently activated system configuration. However I figured its
usefulness to be quite limited as it doesn't have a working
autocompletion and you have to manually type in full option names.

This patch aims to improve the situation by parsing the module system
using `lib.evalModules` and matching the results against the given `$1`.

Internally the `_nix_attr_paths` completion helper is used to provide
proper completion for a given attribute set.

The only known limitation is that option properties (like `default`,
`description`, etc.) are autocompleted as well although they can't be
handled by `nixos-option`. This happens as it can't be reliably decided
whether an attribute name is part of an option or an option itself (e.g.
`boot.loader.grub.default`).

It could be worked around this issue by
extracting all affected arguments using `lib.functionArgs` and filter
all of them if they're part of an option (attribute set with `_type` ==
`option`). However this doesn't apply for all attribute names (e.g.
`declarations` or `definitions`) and not all options properly evaluate
ATM (e.g. `fileSystems`).

Co-authored-by: Tor Hedin Brønner <torhedinbronner@gmail.com>

hedning commented:
Ah, nice. I had basically written this off due to `nixos-option` being pretty slow, assuming that doing it faster wouldn't be doable. But this works really well :)

To solve the _sep_parts issue we should simply use the existing `_nix_attr_paths`. It's the core completion function (which ideally should be easier to figure out) that's used to generate and complete attribute paths.
It takes as its first argument a nix expression describing the unnamed «top level» expression and uses it to generate completions based on the word being completed (taking `-I` options into account). 

Something like this should do:
```
_nix-common-options # import _nix_attr_paths etc.

_nixos-option-opts() {
    local options='((import <nixpkgs/lib>).evalModules { modules = import <nixpkgs/nixos/modules/module-list.nix>; }).options' # top level expression
    _nix_attr_paths $options
}

_arguments \
    $_nix_search_path_args \
    '--xml[Render as XML]' \
    '*:NixOS module options:_nixos-option-opts'

```

Figuring out if we should burrow further down could be nice, though in general it's a bit tricky and can easily cause performance problems. Eg. checking all possible completions for some attribute will most likely just be painfully slow I think.

Ma27 commented:
I updated the completion script (and filtered all attributes starting with `_`).
Also added you as co-author in the commit and documented issues with completion of `default`, defaultText, etc :)

hedning commented:
I changed `*:NixOS ...` to `:NixOS ...` since `nixos-option` only takes a single normal argument. Pushed in 04c1d76dced8d813117d5f4d8073fc5ef3b97578

Thank you, this makes `nixos-option` much more useful :)
